programming-extensions/programming-extension-processbuilder/src/main/resources/processbuilder/linux/kill_process_tree.sh

```diff
@@ -4,12 +4,73 @@
 # pattern. Any process which does not belong the current user will be ingored.
 
 LOGIN=$(whoami)
+
+
+########## DEBUG
+
+#!/bin/bash
+
+trap_with_arg() {
+    func="$1" ; shift
+    for sig ; do
+        trap "$func $sig" "$sig"
+    done
+}
+
+func_trap() {
+    echo Trapped: $1
+}
+
+trap_with_arg func_trap SIGTERM INT EXIT
+
+echo my own pid: $$
+parentPid=`ps -o ppid= -p $$`
+echo parent pid: $parentPid
+############ END DEBUG
+
+######## SEND SIGTERM
+for ppid in `pgrep -u $LOGIN -f $1`;
+do
+  if [ $ppid -ne $$ -a $ppid -ne $parentPid ]
+  then
+    for pid in `pstree -p $ppid | grep -o -E [0-9]+`
+    do
+      echo "Send Sigterm: $pid"
+      kill -15 $pid > /dev/null 2>&1
+    done
+  fi
+done
+
+
+
+###### WAIT FOR PROCESSES TO BE STOPPED
+# Iterate in seconds
+echo "Wait for $(seq 1 $2) seconds"
+for iteration in $(seq 1 $2)
+do
+    echo "Waiting iteration"
+    echo Number of running processes: `pgrep -u $LOGIN -f $1 -c`
+    if [ `pgrep -u $LOGIN -f $1 -c` -eq "2" ] # command_step.sh executes kill_process.tree.sh, so two processes
+    # means nothing else than the killing procedure is running
+    then
+        echo "Break killing loop because processes dissappeared"
+        break # Processes don't exist -> so don't wait
+    fi
+    echo "Sleep a second: $iteration"
+    sleep 1s # Sleep one seconds, because timeout is expressed in seconds
+done
+
+
+
+
+##### SEND SIGKILL AFTER TIMEOUT HIT
 for ppid in `pgrep -u $LOGIN -f $1`;
 do
-  if [ $ppid -ne $$ ]
+  if [ $ppid -ne $$ -a $ppid -ne $parentPid ]
```

Comment:
this succession of for, if, for is very difficult to understand. Could you create a function to make things a bit clearer